File: 20250510-ABC405/d.cpp
#include <iostream>
#include <vector>
#include <deque>
#include <tuple>
#include <string>

using namespace std;

int main() {
    int H, W;
    cin >> H >> W;

    vector<string> grid(H);
    for (int i = 0; i < H; ++i) {
        cin >> grid[i];
    }

    vector<pair<int, int>> dirs = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};
    vector<vector<int>> dist(H, vector<int>(W, -1));
    vector<vector<pair<int, int>>> prev(H, vector<pair<int, int>>(W, {-1, -1}));

    deque<pair<int, int>> q;
    for (int i = 0; i < H; ++i) {
        for (int j = 0; j < W; ++j) {
            if (grid[i][j] == 'E') {
                dist[i][j] = 0;
                q.push_back({i, j});
            }
        }
    }

    while (!q.empty()) {
        int y, x;
        tie(y, x) = q.front();
        q.pop_front();
        for (const auto& dir : dirs) {
            int dy = dir.first, dx = dir.second;
            int ny = y + dy, nx = x + dx;
            if (ny >= 0 && ny < H && nx >= 0 && nx < W) {
                if (grid[ny][nx] == '.' && dist[ny][nx] == -1) {
                    dist[ny][nx] = dist[y][x] + 1;
                    prev[ny][nx] = {-dy, -dx};
                    q.push_back({ny, nx});
                }
            }
        }
    }

    vector<string> res(H, string(W, ' '));
    for (int i = 0; i < H; ++i) {
        for (int j = 0; j < W; ++j) {
            if (grid[i][j] == '#') {
                res[i][j] = '#';
            } else if (grid[i][j] == 'E') {
                res[i][j] = 'E';
            } else {
                int dy = prev[i][j].first, dx = prev[i][j].second;
                for (const auto& dir : dirs) {
                    int ddy = dir.first, ddx = dir.second;
                    if (dy == ddy && dx == ddx) {
                        if (ddy == -1 && ddx == 0) res[i][j] = '^';
                        else if (ddy == 1 && ddx == 0) res[i][j] = 'v';
                        else if (ddy == 0 && ddx == -1) res[i][j] = '<';
                        else if (ddy == 0 && ddx == 1) res[i][j] = '>';
                        break;
                    }
                }
            }
        }
    }

    for (const auto& row : res) {
        cout << row << endl;
    }

    return 0;
}

File: 20250510-ABC405/d.py
import sys
from collections import deque

input = sys.stdin.readline

H, W = map(int, input().split())
grid = [list(input().strip()) for _ in range(H)]

dirs = [(-1, 0, '^'), (1, 0, 'v'), (0, -1, '<'), (0, 1, '>')]

dist = [[-1] * W for _ in range(H)]
prev = [[None] * W for _ in range(H)]

q = deque()
for i in range(H):
    for j in range(W):
        if grid[i][j] == 'E':
            dist[i][j] = 0
            q.append((i, j))

while q:
    y, x = q.popleft()
    for dy, dx, _ in dirs:
        ny, nx = y + dy, x + dx
        if 0 <= ny < H and 0 <= nx < W:
            if grid[ny][nx] == '.' and dist[ny][nx] == -1:
                dist[ny][nx] = dist[y][x] + 1
                prev[ny][nx] = (-dy, -dx)
                q.append((ny, nx))

res = [[''] * W for _ in range(H)]
for i in range(H):
    for j in range(W):
        if grid[i][j] == '#':
            res[i][j] = '#'
        elif grid[i][j] == 'E':
            res[i][j] = 'E'
        else:
            dy, dx = prev[i][j]
            for ddy, ddx, arrow in dirs:
                if dy == ddy and dx == ddx:
                    res[i][j] = arrow
                    break

sys.stdout.write('\n'.join(''.join(row) for row in res) + '\n')
